spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -408,36 +405,42 @@ public void close() throws IOException {
       Iterator<InternalRow> iter;
 
       if (hasJoinedPartitionColumns) {
-        // schema used to read data files
-        Schema readSchema = TypeUtil.selectNot(requiredSchema, idColumns);
         Schema partitionSchema = TypeUtil.select(requiredSchema, idColumns);
-        PartitionRowConverter convertToRow = new PartitionRowConverter(partitionSchema, spec);
-        JoinedRow joined = new JoinedRow();
 
-        InternalRow partition = convertToRow.apply(file.partition());
-        joined.withRight(partition);
+        Map<Integer, Object> partitionValueMap = Maps.newHashMap();
+        Map<Integer, Integer> partitionSpecFieldIndexMap = Maps.newHashMap();
+        Map<Integer, DataType> partitionSpecDataTypeMap = Maps.newHashMap();
+        for (int i = 0; i < spec.fields().size(); i++) {
+          Integer sourceId = spec.fields().get(i).sourceId();
+          partitionSpecFieldIndexMap.put(sourceId, i);
+          partitionSpecDataTypeMap.put(sourceId, getPartitionType(partitionSchema, sourceId));
```

Comment:
Why does this use `partitionSchema` instead of `spec.partitionType()`? The partition schema is made from the partition columns in the table schema, which could be out of order. The spec's type is the struct type returned by `file.partition()`.

Reply:
👍